File: core/src/main/java/feast/core/model/Source.java
/*
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2018-2019 The Feast Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package feast.core.model;

import com.google.protobuf.TextFormat;
import feast.proto.core.SourceProto;
import feast.proto.core.SourceProto.KafkaSourceConfig;
import feast.proto.core.SourceProto.Source.Builder;
import feast.proto.core.SourceProto.SourceType;
import io.grpc.Status;
import java.util.Objects;
import javax.persistence.*;
import javax.persistence.Entity;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;

@Entity
@Getter
@Setter
@AllArgsConstructor
@Table(name = "sources")
public class Source {

  /** Source Id. Internal use only, do not use to identify the source. */
  @Id
  @GeneratedValue
  @Column(name = "pk")
  private Integer id;

  @Deprecated
  @Column(name = "id")
  private String deprecatedId;

  @Deprecated
  @Column(name = "bootstrap_servers")
  private String bootstrapServers;

  @Deprecated
  @Column(name = "topics")
  private String topics;

  /** Type of the source */
  @Enumerated(EnumType.STRING)
  @Column(name = "type", nullable = false)
  private SourceType type;

  /** Configuration object specific to each source type */
  @Column(name = "config")
  private String config;

  @Column(name = "is_default")
  private boolean isDefault;

  public Source() {
    super();
  }

  public String getConfig() {
    if ((config == null || config.isEmpty()) && bootstrapServers != null && topics != null) {
      config =
          KafkaSourceConfig.newBuilder()
              .setBootstrapServers(bootstrapServers)
              .setTopic(topics)
              .build()
              .toString();
    }

    return config;
  }

  /**
   * Construct a source facade object from a given proto object.
   *
   * @param sourceSpec SourceProto.Source object
   * @param isDefault Whether to return the default source object if the source was not defined by
   *     the user
   * @return Source facade object
   */
  public static Source fromProto(SourceProto.Source sourceSpec, boolean isDefault) {

    if (sourceSpec.equals(SourceProto.Source.getDefaultInstance())) {
      Source source = new Source();
      source.setDefault(true);
      return source;
    }

    Source source = new Source();
    source.setType(sourceSpec.getType());

    switch (sourceSpec.getType()) {
      case KAFKA:
        if (sourceSpec.getKafkaSourceConfig().getBootstrapServers().isEmpty()
            || sourceSpec.getKafkaSourceConfig().getTopic().isEmpty()) {
          throw Status.INVALID_ARGUMENT
              .withDescription(
                  "Unsupported source options. Kafka source requires bootstrap servers and topic to be specified.")
              .asRuntimeException();
        }
        source.setConfig(sourceSpec.getKafkaSourceConfig().toString());
        break;
      case UNRECOGNIZED:
      default:
        throw Status.INVALID_ARGUMENT
            .withDescription("Unsupported source type. Only [KAFKA] is supported.")
            .asRuntimeException();
    }

    source.setDefault(isDefault);
    return source;
  }

  /**
   * Construct a source facade object from a given proto object.
   *
   * @param sourceSpec SourceProto.Source object
   * @return Source facade object
   */
  public static Source fromProto(SourceProto.Source sourceSpec) {
    return fromProto(sourceSpec, false);
  }

  /**
   * Convert this object to its equivalent proto object.
   *
   * @return SourceProto.Source
   */
  public SourceProto.Source toProto() {
    Builder builder = SourceProto.Source.newBuilder().setType(this.getType());

    switch (this.getType()) {
      case KAFKA:
        KafkaSourceConfig.Builder kafkaSourceConfig = KafkaSourceConfig.newBuilder();
        try {
          com.google.protobuf.TextFormat.getParser().merge(this.getConfig(), kafkaSourceConfig);
        } catch (TextFormat.ParseException e) {
          throw new RuntimeException(
              String.format(
                  "Unable to deserialize source configuration from String to KafkaSourceConfig: %s",
                  this.getConfig()),
              e);
        }
        return builder.setKafkaSourceConfig(kafkaSourceConfig).build();
      case INVALID:
      case UNRECOGNIZED:
      default:
        throw new RuntimeException(
            String.format(
                "Unable to build Source from configuration and type: %s %s",
                this.getConfig(), this.getType()));
    }
  }

  /**
   * Override equality for sources. Sources are compared based on their type and type-specific
   * options.
   *
   * @param other other Source
   * @return boolean equal
   */
  public boolean equalTo(Source other) {
    if ((this.getType() == null || !this.getType().equals(other.getType()))) {
      return false;
    }

    return this.getConfig().equals(other.getConfig());
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    Source source = (Source) o;
    return this.equalTo(source);
  }

  @Override
  public int hashCode() {
    return Objects.hash(getType(), getConfig());
  }

  /**
   * Returns the type of this Source in String format
   *
   * @return Source type in String format
   */
  public String getTypeString() {
    return this.getType().getValueDescriptor().getName();
  }
}

File: sdk/python/feast/source.py
# Copyright 2019 The Feast Authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from feast.core.Source_pb2 import (
    Source as SourceProto,
    KafkaSourceConfig as KafkaSourceConfigProto,
    SourceType as SourceTypeProto,
)


class Source:
    """
    Source is the top level class that represents a data source for finding
    feature data. Source must be extended with specific implementations to
    be useful
    """

    def __eq__(self, other):
        return True

    @property
    def source_type(self) -> str:
        """
        The type of source. If not implemented, this will return "None"
        """
        return "None"

    def to_proto(self):
        """
        Converts this source object to its protobuf representation.
        """
        return None

    @classmethod
    def from_proto(cls, source_proto: SourceProto):
        """
        Creates a source from a protobuf representation. This will instantiate
        and return a specific source type, depending on the protobuf that is
        passed in.

        Args:
            source_proto: SourceProto python object

        Returns:
            Source object
        """
        if source_proto.type == SourceTypeProto.KAFKA:
            return KafkaSource(
                brokers=source_proto.kafka_source_config.bootstrap_servers,
                topic=source_proto.kafka_source_config.topic,
            )

        return cls()


class KafkaSource(Source):
    """
    Kafka feature set source type.
    """

    def __init__(self, brokers: str = "", topic: str = ""):
        """

        Args:
            brokers: Comma separated list of Kafka brokers/bootstrap server
                addresses, for example: my-host:9092,other-host:9092
            topic: Kafka topic to find feature rows for this feature set
        """
        self._source_type = "Kafka"
        self._brokers = brokers
        self._topic = topic

    def __eq__(self, other):
        if (
            self.brokers != other.brokers
            or self.topic != other.topic
            or self.source_type != other.source_type
        ):
            return False
        return True

    @property
    def brokers(self) -> str:
        """
        Returns the list of broker addresses for this Kafka source
        """
        return self._brokers

    @property
    def topic(self) -> str:
        """
        Returns the topic for this feature set
        """
        return self._topic

    @property
    def source_type(self) -> str:
        """
        Returns the type of source. For a Kafka source this will always return
            "kafka"
        """
        return self._source_type

    def to_proto(self) -> SourceProto:
        """
        Converts this Source into its protobuf representation
        """
        return SourceProto(
            type=SourceTypeProto.KAFKA,
            kafka_source_config=KafkaSourceConfigProto(
                bootstrap_servers=self.brokers, topic=self.topic
            ),
        )
